content/blog/2018-10-09-jstor.md

```diff
@@ -0,0 +1,512 @@
+  old_approach()
+})
+
+#> Error in parse_rprof(prof_output, expr_source) : No parsing data available. 
+#> Maybe your function was too fast?
+```
+
+A simple solution is to rerun the function many times. Note, however, that this
+does not increase reliability and reproducibility of the results. The function is
+not measured 500 times, the result
+being the mean or median (like in `microbenchmark`), but it is simply the 
+aggregate of running the function 500 times. This can vary quite a bit. All in
+all, tough, I found it still useful to judge if any part of the code is orders 
+of magnitude slower than the rest.
+
+In the following chunk
+I define the function again for two reasons: first to separate all computations
+into separate lines. This ensures, that we get a measurement for each line.
+Second, the code needs to be supplied either within the call to `profvis`, or it
+must be defined in a sourced file, otherwise the output will not be as 
+informative, because `profvis` will not be able to access the functions' source
+code.
+
+
+```r
+profvis::profvis({
+  
+  old_approach <- function(authors) {
+    res <- tibble::data_frame(id = "1",
+                              authors = list(authors))
+    
+    tidyr::unnest(res, authors)
+  }
+
+  purrr::rerun(500, old_approach(authors))
+})
+```
+
+![](/img/blog-images/2018-10-09-jstor/old_version.png)
+
+We can see that the call to `tidyr::unnest` takes up a lot of time, thus
+it would be good to get rid of it. The new approach does not need unnesting, and
+is roughly 17 times faster:
+
+
+
+```r
+profvis::profvis({
+
+  new_approach <- function(authors) {
+    data.frame(id = "1",
+               authors = authors)
+  }
+  
+  purrr::rerun(500, new_approach(authors))})
+```
+
+![](/img/blog-images/2018-10-09-jstor/new_version.png)
+
+In the end, by assessing the efficiency of functions repeatedly and optimising several
+parts,
+I was able to trim down execution time considerably overall. For 25,000 files,
+which is the maximum number of files you can receive at one time through the
+standard interface of JSTOR/DfR, execution time is slightly under 4 minutes, or
+2 minutes if executed in parallel (at least on my machine with a 2,8GHz Intel 
+i5).
+
+
+### Lessons learned
+I have learned many things while working on this package. While I acquired 
+certain skills (like writing simple XPATH-queries), I want to highlight a few 
+general things.
+
+Something which is probably true for many people working with and developing for
+R, is that community is important. Without the efforts of many others, through
+paid work or by spending their time voluntarily, developing the package would
+not have been possible. This is true for the many packages my code builds on,
+for the suite of packages that helps in developing and maintaining packages,
+for answers I got over StackOverflow, and last but not least for the helpful
+comments I received by @elinw and @jsonbecker during the review process for
```

Comment:
Thank you for acknowledging the reviewers! Please link to them like:
`[Elin Waring](https://github.com/elinw) and [Jason Becker](https://github.com/jsonbecker)`

Reply:
If you feel comfortable with it you could link to the open review like `[review process](https://github.com/ropensci/onboarding/issues/189)`. Given the emphasis of this post about your process and improvements, that might be helpful for readers to get insight into the review process.

Reply:
Thanks for both of these hints. These are actually stated in the guidelines and I missed them – sorry!